### ISSUE TITLE
Slow Async GPU Copy

### ISSUE THREAD

**Issue by stephenrawls** [apache/incubator-mxnet#14210]
Some of our NLP models need to use multiple data arrays, sometimes up to 6 different data ndarrays as input. This is a big problem for multi-gpu training because the CPU -> GPU copy time is very slow. In particular, whereas the *actual* time to copy data via cuda memcpy calls is fast, the python overhead with calling into the MxNet C API seems slow.

Here is a sample script that shows what I mean:
```
import time
import mxnet as mx

def generate_data(num_inputs, ngpus):
    return[[mx.nd.random.randn(1) for _ in range(num_inputs)] for _ in range(ngpus)]

def send_to_gpu(data, ctx_list):
    return [[data[i][j].as_in_context(ctx_list[i]) for j in range(len(data[0]))] for i in range(len(data))]

ctx_list = [mx.gpu(i) for i in range(8)]

# Need to do this as warmup                                                                                                                                                                       
data = generate_data(num_inputs=1, ngpus=8)
send_to_gpu(data, ctx_list)


for num_inputs in range(1,11):
    data = generate_data(num_inputs, ngpus=8)

    start = time.time()
    data = send_to_gpu(data, ctx_list)
    end = time.time()

    print("Num Inputs: %d. Took %f ms to set off all async copies" % (num_inputs, 1000*(end-start)))
```

The output on a p3.16xlarge instance is:
```
% python3 ~/test_send_to_gpu_gh.py
Num Inputs: 1. Took 0.680685 ms to set off all async copies
Num Inputs: 2. Took 2.131939 ms to set off all async copies
Num Inputs: 3. Took 2.979040 ms to set off all async copies
Num Inputs: 4. Took 4.072189 ms to set off all async copies
Num Inputs: 5. Took 4.901409 ms to set off all async copies
Num Inputs: 6. Took 7.693768 ms to set off all async copies
Num Inputs: 7. Took 6.579638 ms to set off all async copies
Num Inputs: 8. Took 7.871866 ms to set off all async copies
Num Inputs: 9. Took 8.602858 ms to set off all async copies
Num Inputs: 10. Took 9.772539 ms to set off all async copies
```

My thoughts are: 

(1) Can I move this overhead out of the main training thread and hide the latency the same way we hide data loading latency? I think to do that I would need something like the cuda IPC support that PyTorch has: https://github.com/pytorch/pytorch/blob/220ce8046e5fcf1434f948795bcdefda33e95e9a/torch/multiprocessing/reductions.py

(2) Can I reduce the overhead associated with each `.as_in_context()` call, which I am *mostly* sure is an async call and is just suffering from high overhead. My thought there was to try the cython support, which at least some places on the internet suggest will have lower overhead when calling into the C API, but it looks like that is currently broken pending this patch: https://github.com/apache/incubator-mxnet/pull/10951

Can someone look at this and let me know if I'm doing anything obviously wrong with the way I am trying to asynchronously copy data to each GPU? And is there an easy way to get faster CPU -> GPU copy times?  (Again, the actual data is relatively small and doing the copy is quick; the problem is the overhead from needing to call copy 8 (gpus) * 6 (num input arrays) = 48 times)

Thanks!

**Comment by stephenrawls** [comment]
As a follow up, I decided to time how long a single call into the c api takes for a no-op style operation:

```
import time
import mxnet as mx
import numpy as np

x = mx.nd.array([0])

# warmup                                                                                                                                                                                          
y = mx.nd.identity(x)


times = []
for _ in range(1000):
    start = time.time()
    y = mx.nd.identity(x)
    end = time.time()
    times.append(1000 * (end-start))


print("Took %0.3f  +/- %0.2fms" % (np.mean(times), np.std(times)))
```

The output I get is: 
```
Took 0.037  +/- 0.01ms
```

If I multiply this by 48 then I get an expected overhead of 1.776 ms. Which is still too slow to pay at the start of every training loop, but there is a big gap between that and the 6-7 ms that I am observing.

It turns out the `.as_in_context` is also calling the following code to create an nd array handle:
```
NDArray(_new_alloc_handle(self.shape, other, True, self.dtype))
```

From my timing tests this adds another 0.816 ms of latency when doing 48 calls.

So I guess what we have is with 8 GPUs and 6 input arrays per GPU, roughly:
```
1.777 ms latency from call into c api
0.816 ms from new ndarray handle allocation
3-5 ms from, I guess CopyFromToSimple() & its call graph
```

**Comment by frankfliu** [comment]
@mxnet-label-bot add [performance, python]

**Comment by eric-haibin-lin** [comment]
How many NDArrays are you copying? I know it is not a good answer for this issue but usually I find it fast enough to prefetch the next batch of ndarrays to GPU ahead of time, while the GPU is busy computing the current batch.

**Comment by stephenrawls** [comment]
@eric-haibin-lin  I would be okay with that solution, but not sure the best way to achieve it.

Currently my training loop looks roughly like this:
```
for data in data_loader:
  data_on_gpus = send_data_to_gpus(data)
  losses = train_iteration(data_on_gpus)
  loss = sum([loss.asscalar() for loss in losses]) / len(losses)
```

This is just rough idea of course, but the point is that the data loader is happening on separate processes so the latency for getting data is hidden. But I am waiting at the bottom of my training loop to be able to read the loss, meaning that at the top of the training loop I have to copy data to gpu before sending data to model.

I would like to be able to have the data loader just put the data on the gpu directly, but for this to be the case I'm pretty sure we would need IPC communication for cuda ndarrays. PyTorch has this support but it doesn't look like MXNet does.

I guess I could try to re-jigger my training loop so that I don't have to wait on the loss from the previous training iteration until after the call to send_data_to_gpus(). That will make my code a bit more messy but maybe be a sufficient work around.

Thanks for input. Is there any way to reduce the overhead of calling into the c api? Maybe create a new operator that accepts a list of ndarrays and does `.as_in_context(gpu)` for all of them in one go?

**Comment by eric-haibin-lin** [comment]
That is certainly sth to try, although I am not sure why `CopyFromToSimple` takes 3-5 ms for 48 runs. I thought `CopyFromToSimple` allocates memory but it turns out not the case. The actual memory allocation seems to be inside copy<cpu, gpu> when TBlob() is invoked. https://github.com/apache/incubator-mxnet/blob/5f32f32e8c7ce50f20eae1438bf970eba6830578/src/ndarray/ndarray.cc#L1186-L1275 
Did you profile which part of `CopyFromToSimple` call graph is slow?
